From 09bac5ed1b9a94c29d14cae7bec12ba3f2589f78 Mon Sep 17 00:00:00 2001
From: Johnny Mnemonic <jm@machine-hall.org>
Date: Sat, 6 Dec 2025 15:41:07 +0100
Subject: [PATCH 95/96] ia64: Disable non-working check

Since 7203ca412fc8e8a0588e9adc0f777d3163f8dff3 there's a build regression
present:
```
  CC      arch/ia64/mm/init.o
  AR      init/built-in.a
  CC      kernel/sched/build_policy.o
  CC      arch/ia64/kernel/ptrace.o
  CC      arch/ia64/mm/fault.o
In file included from ./include/linux/bits.h:5,
                 from ./include/linux/bitops.h:6,
                 from ./include/linux/log2.h:12,
                 from ./include/asm-generic/getorder.h:8,
                 from ./arch/ia64/include/asm/page.h:119,
                 from ./arch/ia64/include/asm/ptrace.h:46,
                 from ./arch/ia64/include/asm/processor.h:20,
                 from ./include/linux/sched.h:13,
                 from ./include/linux/rcupdate.h:27,
                 from ./include/linux/rculist.h:11,
                 from ./include/linux/sched/signal.h:5,
                 from arch/ia64/mm/fault.c:8:
arch/ia64/mm/fault.c: In function 'ia64_do_page_fault':
./include/linux/mm.h:398:41: warning: "int" is not defined, evaluates to 0 [-Wundef]
  398 | #define INIT_VM_FLAG(name) BIT((__force int) VMA_ ## name ## _BIT)
      |                                         ^~~
./include/vdso/bits.h:7:44: note: in definition of macro 'BIT'
    7 | #define BIT(nr)                 (UL(1) << (nr))
      |                                            ^~
./include/linux/mm.h:399:25: note: in expansion of macro 'INIT_VM_FLAG'
  399 | #define VM_READ         INIT_VM_FLAG(READ)
      |                         ^~~~~~~~~~~~
arch/ia64/mm/fault.c:126:36: note: in expansion of macro 'VM_READ'
  126 | #       if (((1 << VM_READ_BIT) != VM_READ || (1 << VM_WRITE_BIT) != VM_WRITE) \
      |                                    ^~~~~~~
./include/linux/mm.h:398:46: error: missing binary operator before token "VMA_READ_BIT"
  398 | #define INIT_VM_FLAG(name) BIT((__force int) VMA_ ## name ## _BIT)
      |                                              ^~~~
./include/vdso/bits.h:7:44: note: in definition of macro 'BIT'
    7 | #define BIT(nr)                 (UL(1) << (nr))
      |                                            ^~
./include/linux/mm.h:399:25: note: in expansion of macro 'INIT_VM_FLAG'
  399 | #define VM_READ         INIT_VM_FLAG(READ)
      |                         ^~~~~~~~~~~~
arch/ia64/mm/fault.c:126:36: note: in expansion of macro 'VM_READ'
  126 | #       if (((1 << VM_READ_BIT) != VM_READ || (1 << VM_WRITE_BIT) != VM_WRITE) \
      |                                    ^~~~~~~
make[6]: *** [scripts/Makefile.build:287: arch/ia64/mm/fault.o] Error 1
make[5]: *** [scripts/Makefile.build:556: arch/ia64/mm] Error 2
```

The definition of VM_READ was changed from `0x00000001` (direct value) to
`INIT_VM_FLAG(READ)` (macro call).

Looking into arch/ia64/mm/init.c where the same definition of VM_READ can
be expanded w/o problems everywhere it's used, it looks like the expansion
does not work correctly when VM_READ is used like in arch/ia64/mm/fault.c
on line 126. Not sure if this is a preprocessor limitation, but this would
also affect VM_WRITE and VM_EXEC in the same way then.

As the actual check can't work anyhow then, let's disable it for now until
a better solution can be found and on the way workaround the build
regression.

Tested to work in Ski and also for a boot and small compile job on a
rx2620.
---
 arch/ia64/mm/fault.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/ia64/mm/fault.c b/arch/ia64/mm/fault.c
index 9024a29207b1..32d186454bdc 100644
--- a/arch/ia64/mm/fault.c
+++ b/arch/ia64/mm/fault.c
@@ -123,10 +123,12 @@ ia64_do_page_fault (unsigned long address, unsigned long isr, struct pt_regs *re
 
 	/* OK, we've got a good vm_area for this memory area.  Check the access permissions: */
 
+#if 0
 #	if (((1 << VM_READ_BIT) != VM_READ || (1 << VM_WRITE_BIT) != VM_WRITE) \
 	    || (1 << VM_EXEC_BIT) != VM_EXEC)
 #		error File is out of sync with <linux/mm.h>.  Please update.
 #	endif
+#endif
 
 	if (((isr >> IA64_ISR_R_BIT) & 1UL) && (!(vma->vm_flags & (VM_READ | VM_WRITE))))
 		goto bad_area;
-- 
2.25.1

